### PR TITLE
(feat) Support code-scanning API (Pt#1)

### PR DESCRIPTION
**Pull request by akashRindhe**
# Description

Related issue - https://github.com/hub4j/github-api/issues/1133
New feature supporting GitHub code-scanning API. 

As this is a big feature with multiple functions, breaking this into smaller PRs to reduce review effort. The functions supported in this PR are -

- [Listing all code-scanning alerts for a repository](https://docs.github.com/en/rest/reference/code-scanning#list-code-scanning-alerts-for-a-repository)
- [Get a code-scanning alert for a repository by its id](https://docs.github.com/en/rest/reference/code-scanning#get-a-code-scanning-alert)
- [List all instances for a code-scanning alert](https://docs.github.com/en/rest/reference/code-scanning#list-instances-of-a-code-scanning-alert)

### CODE REVIEW

**Comment by bitwiseman** [comment]
@akashRindhe 
We value contributions like this.  I apologize for the delay reviewing and merging.  We've seen some high priority issues come in and I personally have other work I've needed to focus on.  I will get to this as soon as I can.

File: src/main/java/org/kohsuke/github/GHCodeScanningAlert.java [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+        return rule;
+    }
+
+    /**
+     * Most recent instance of the alert
+     *
+     * @return most recent instance
+     */
+    public GHCodeScanningAlertInstance getMostRecentInstance() {
+        return most_recent_instance;
```
Comment (bitwiseman) [inline src/main/java/org/kohsuke/github/GHCodeScanningAlert.java:128]: @akashRindhe 
This is marked as not covered.  Please see if there's anything you can do to cover this method and any other methods that are not covered by tests.  Don't wear yourself out on it, but just take another look and see if any of them are easy cover.

**Review by bitwiseman — CHANGES_REQUESTED** [review]
Please review all messages regarding lines not covered from code coverage and fix or comment on why they are not covered.

**Comment by bitwiseman** [comment]
@ravingbool 
Do you think you could do the testing to get this in shape to merge?  (You reacted 👍 above, so I thought it might be of interest to you.)

File: src/main/java/org/kohsuke/github/GHRepository.java [hunk 2]
```diff
@@ -3178,6 +3178,58 @@ public GHTagObject createTag(String tag, String message, String object, String t
...
+    /**
+     * Lists the code scanning alerts of this repository filtered on the code scanning tool name
+     *
+     * @param toolName
+     *            name of code scanning tool that creates alerts
+     * @return the paged iterable
+     */
+    public PagedIterable<GHCodeScanningAlert> listCodeScanningAlerts(String toolName) {
+        return listCodeScanningAlerts(Collections.singletonMap("tool_name", toolName));
+    }
```
Comment (bitwiseman) [inline src/main/java/org/kohsuke/github/GHRepository.java:3210]: There are other filters: 
https://docs.github.com/en/rest/reference/code-scanning#list-code-scanning-alerts-for-a-repository

These filters can be combined so having only single filter method available doesn't seem right.

**Comment by bitwiseman** [comment]
Converted to Draft as it is not ready for review.

**Comment by jgangemi** [comment]
hey @bitwiseman / @akashRindhe where does this stand? i could use this functionality at the day job and am happy to pick up on it.

**Comment by bitwiseman** [comment]
> hey @bitwiseman / @akashRindhe where does this stand? i could use this functionality at the day job and am happy to pick up on it.

What's needed is a review of all messages regarding lines not covered from code coverage and fix or comment on why they are not covered.

**Comment by jgangemi** [comment]
ok - i've created a new branch to work on. i'll push something up for some initial feedback.

**Comment by bitwiseman** [comment]
> ok - i've created a new branch to work on. i'll push something up for some initial feedback.

Just open a new PR. That will get you all the checking behavior that you're going to need.  Just keep @akashRindhe's original commits and build on top of them. That way they still get the credit for their work when the PR is merged.

**Pull request closed without merging**